bus/contractlocking.go

```diff
@@ -0,0 +1,209 @@
+	lock.wakeupTimer = time.AfterFunc(d, func() {
+		l.Release(id, lockID)
+	})
+}
+
+func (l *contractLock) stopTimer() {
+	if l.wakeupTimer == nil {
+		return
+	}
+	if !l.wakeupTimer.Stop() {
+		select {
+		case <-l.wakeupTimer.C:
+		default:
+		}
+	}
+}
+
+// Acquire acquires a contract lock for the given id and provided duration. If
+// acquiring the lock doesn't finish before the context is closed,
+// ErrAcquireContractTimeout is returned. Upon success an identifier is returned
+// which can be used to release the lock before its lock duration has passed.
+// TODO: Extend this with some sort of priority. e.g. migrations would acquire a
+// lock with a low priority but contract maintenance would have a very high one
+// to avoid being starved by low prio tasks.
+func (l *contractLocks) Acquire(ctx context.Context, priority int, id types.FileContractID, d time.Duration) (uint64, error) {
+	lock := l.lockForContractID(id, true)
+
+	// Prepare a random lockID for ourselves.
+	ourLockID := frand.Uint64n(math.MaxUint64) + 1
+
+	lock.mu.Lock()
+
+	// If nobody holds the lock, acquire it and launch a timer to release
+	// the lock after the expiry.
+	if lock.heldByID == 0 {
+		lock.heldByID = ourLockID
+		lock.setTimer(l, ourLockID, id, d)
+		lock.mu.Unlock()
+		return ourLockID, nil
+	}
+
+	// Someone is holding the lock. Add ourselves to the queue.
+	wakeChan := make(chan struct{})
+	heap.Push(lock.queue, &lockCandidate{
+		wake:     wakeChan,
+		priority: priority,
+		timedOut: ctx.Done(),
+	})
+
+	lock.mu.Unlock()
+	select {
+	case <-ctx.Done():
+		// NOTE: We need to acquire the lock again and check for a
+		// wakeup. That way we close a gap where a thread could time out
+		// while being woken up.
+		lock.mu.Lock()
+		select {
+		case <-wakeChan:
+		default:
+			lock.mu.Unlock()
+			return 0, ErrAcquireContractTimeout
+		}
+	case <-wakeChan:
+		lock.mu.Lock()
+	}
+
+	if lock.heldByID != 0 {
+		panic("lock should be released after being woken up")
+	}
+	lock.heldByID = ourLockID
+	lock.setTimer(l, ourLockID, id, d)
+	lock.mu.Unlock()
+	return ourLockID, nil
+}
+
+// Release releases the contract lock for a given contract and lock id.
+func (l *contractLocks) Release(id types.FileContractID, lockID uint64) error {
+	if lockID == 0 {
+		return errors.New("can't release lock with id 0")
+	}
+	lock := l.lockForContractID(id, false)
+	if lock == nil {
+		return nil // nothing to do
+	}
+
+	lock.mu.Lock()
+	defer lock.mu.Unlock()
+	if lock.heldByID == 0 {
+		return nil // nothing to do
+	}
+	if lock.heldByID != lockID {
+		return fmt.Errorf("can't unlock lock due to id mismatch %v != %v", lockID, lock.heldByID)
+	}
+
+	// Stop the timer on the lock.
+	lock.stopTimer()
+
+	// Set holder to 0.
+	lock.heldByID = 0
+
+	// If there is no next candidate we are done.
+	if lock.queue.Len() == 0 {
+		return nil
+	}
+
+	// Wake the next candidate.
+	for next := heap.Pop(lock.queue).(*lockCandidate); next != nil; next = heap.Pop(lock.queue).(*lockCandidate) {
```

Comment:
why are you popping it twice? if that's a mistake why is the unit test passing :p

Reply:
guess you meant to peek

Reply:
oh boy nevermind